### PR TITLE
Filetree

### PR DESCRIPTION
**Pull request by ilyafyodorov**
Filetree is implemented with "Populating the tree using a callback function". While it is not really necessary, I decided to keep Model for Filetree (file filetree.js) in case if we'll need to store some addditional data related to files and directories (e.g. for search).

Please note that "MODEL/" is added into the path. It is not relevant now but will be necessary in furure to separate requests for model display using a direct link from the other server requests.

This version is now running at 
https://evening-eyrie-59340.herokuapp.com/

I've tested it on iPhone and it works fine (now you do not have to click on the small triangle to open a directory; you can click on the directory name as well).
There is still a problem to select result on the iphone but it is not related to this branch.

Missing things:
- refresh of the already opened filetree if directories are changed on the server. It can be done in several ways, but I think it is a low proprity.

### CODE REVIEW

**Comment by STYTZ** [comment]
Good job! I'll have a look through the code, see further comments. Also, we need to check what kind of merge conflicts we have there. It could be difficult to resolve, some of your lates commits were quite messy.

**Comment by ilyafyodorov** [comment]
just compare the final commit
https://github.com/STYTZ/MI-web-tool/compare/master...filetree
it is very straightforward

In short, to make it work with our current Master you just need to insert latest VIewport implementation (modules/viewport).

**Comment by STYTZ** [comment]
Yep, you're right, merge should look ok.

File: app/js/modules/navbar/navbar.js [hunk 1]
```diff
@@ -16,16 +20,96 @@ define(function (require) {
...
-            //initialization of the graphical tree using "jstree" library
             $(this.el).jstree({
-                core: {
-                    data: this.model.jsonData
+                'core' : {
+                    //definition of jstree contents via "callback function" method - see jstree documentation
+                    'data' : function (obj, cb) {
+                        
+                        // if root directory is requested
+                        if(obj.id === "#") {
```
Comment (STYTZ) [inline app/js/modules/navbar/navbar.js]: Why root node is treated separately?

**Comment by STYTZ** [comment]
Regarding **filetree.js** - as I understood it is not used at all at the moment, and actually I don't see a reason to use it in near future. So for now I would delete it for clarity.

File: app/js/modules/navbar/navbar.js [hunk 2]
```diff
@@ -34,19 +118,22 @@ define(function (require) {
                         "icon" : "glyphicon glyphicon-th-large"
                     },
                     'dirS' : {
-                        "icon" : "glyphicon glyphicon-th"
+                        "icon" : "glyphicon glyphicon-th-large"
```
Comment (STYTZ) [inline app/js/modules/navbar/navbar.js]: So,**dirS** type is not used? Why not get rid of it?

File: app/js/modules/navbar/navbar.js [hunk 3]
```diff
@@ -1,21 +1,183 @@
...
+                
+                 // plugin usage: "types" to be able to define different types of data in the tree (file/folder etc.)
+                plugins : [ "types" ],    
+      
+                types : {
+                    //3 types defined: dir - directory, dirS - scanned directory (contents were already read), obj - file
+                    'dir' : {
+                        "icon" : "glyphicon glyphicon-th-large"
+                    },
+                    'dirS' : {
```
Comment (STYTZ) [inline app/js/modules/navbar/navbar.js]: So, **dirS** type is not used? Why not get rid of it?

File: app/index.html [hunk 4]
```diff
@@ -30,11 +31,30 @@
...
 
       </div>
     </nav>
+    
+      <!-- Modal: Model Database Browse -->
+    <div class="modal fade" id="browseModal" >
+        <div class="modal-dialog" id="BrowseModalDialog">
+         <div class="modal-content" id="BrowseModalContent">
+            <div class="modal-header">
+              <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
```
Comment (STYTZ) [inline app/index.html]: First "close" button.

File: app/index.html [hunk 5]
```diff
@@ -30,11 +31,30 @@
...
+              <h4 class="modal-title" id="myModalLabel">Browse model database</h4>
+            </div>
+            <div class="modal-body">
+                <div id="jstree-tree"></div>
+            </div>
+            <div class="modal-footer">
```
Comment (STYTZ) [inline app/index.html]: Second "close button". Do we need both?

Comment (ilyafyodorov) [inline app/index.html]: Having 2 buttons for closing a file open dialog is a typical practice, e.g.:
http://www.geekpedia.com/pics/openFileDialog/openFileDialog.gif
Of course it is not strictly needed but it is quite standard.

File: app/js/modules/navbar/navbar.js (hunk 3 above)
Comment (ilyafyodorov) [inline app/js/modules/navbar/navbar.js]: I still mark a directory with "dirS" after its contents are read. Currently it is not used, but perhaps in future we will need to know if a directory was already scanned or not. Maybe during search

File: app/js/modules/navbar/navbar.js (hunk 1 above)
Comment (ilyafyodorov) [inline app/js/modules/navbar/navbar.js]: jstree may ask for root node with "#" request or for some directory/file with actual name of this directory/file. Maybe this difference can be taken into account in some clever fashion but why bother? :)

File: app/js/modules/navbar/navbar.js (hunk 3 above)
Comment (STYTZ) [inline app/js/modules/navbar/navbar.js]: Jstree has **is_loaded()** method for that purpose, **type** is needed for icons only.

Have you seen the *jstree.search* plugin BTW?

**Comment by ilyafyodorov** [comment]
Merge conflicts were resolved, now it can be merged automatically

File: app/js/modules/navbar/navbar.js (hunk 3 above)
Comment (ilyafyodorov) [inline app/js/modules/navbar/navbar.js]: jstree.search serches in alredy loaded tree. But for us, search would have to be implemented on the server (we would search for file or folder names) and then the tree should somehow be built with the server response.

Comment (STYTZ) [inline app/js/modules/navbar/navbar.js]: https://www.jstree.com/api/#/?q=search.ajax
That would be really surprising if jstree couldn't search on the server.

Comment (ilyafyodorov) [inline app/js/modules/navbar/navbar.js]: Interesting! We'll see how it works when we come to search implementation.

File: app/js/modules/navbar/navbar.js (hunk 1 above)
Comment (STYTZ) [inline app/js/modules/navbar/navbar.js]: I really don't get what's the difference. I'll have a closer look later when I get home.

File: app/js/modules/navbar/navbar.js [hunk 6]
```diff
@@ -28,56 +28,32 @@ define(function (require) {
...
-                                                    jsonData.push({"text" : parseObj.val, "id": counter+1, "children" : haveChildren, "type" : parseObj.cls, "state" : {selected  : false}});
-
-                                                    counter = counter+1;
-                                               }
-                                        }
-                                        
-                                        cb(jsonData);
-                                        
-                                }
+                            xhr.open('GET', this.url, true);
```
Comment (STYTZ) [inline app/js/modules/navbar/navbar.js]: Thanks, it's much more clear now!

File: app/js/modules/navbar/navbar.js [hunk 7]
```diff
@@ -1,22 +1,150 @@
...
+                    //definition of jstree contents via "callback function" method - see jstree documentation
+                    'data' : function (obj, cb) {
+                        
+                        //get the url of data root (defined in main.js)            
+                        this.url = radio.request("data:root");
+
+                        var xhr = new XMLHttpRequest();
+                        
+
+                        if(obj.id === "#") {
```
Comment (STYTZ) [inline app/js/modules/navbar/navbar.js]: It looks much more clear now, thanks!

File: app/js/modules/navbar/navbar.js [hunk 8]
```diff
@@ -1,22 +1,150 @@
...
+            this.manipulate(this);
+
+
+        },   
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+        //this function manitulates the view of the file tree and performes fetching of selected directories and files
+        //it is called from corresponding view (navbar.js) after its initialization
+        manipulate: function (view) {
```
Comment (STYTZ) [inline app/js/modules/navbar/navbar.js]: If you don't mind, I'd like to improve this one a tiny bit before merging the pull request. I'll commit some changes later today.

Comment (ilyafyodorov) [inline app/js/modules/navbar/navbar.js]: Sure!

**Pull request merged**